### PR TITLE
[PT TE] Support aten::index_copy_ operation

### LINKED ISSUE #21970: [Good First Issue]: Support aten::index_copy_ operation
### Context

OpenVINO component responsible for support of PyTorch models is called as PyTorch Frontend (PT FE). PT FE converts a model represented as TorchScript model to a model in [OpenVINO opset](https://docs.openvino.ai/latest/openvino_docs_ops_opset.html).

### What needs to be done?

This operation can be found in `transfo-xl-wt103` model from [huggingface](https://huggingface.co/transfo-xl-wt103). You can reproduce the issue with the following code:

```python
import inspect
import os
import torch
import openvino as ov
from openvino.frontend.pytorch.utils import prepare_example_inputs_and_model
from transformers import TransfoXLTokenizer, TransfoXLLMHeadModel

os.environ["TRUST_REMOTE_CODE"] = "True"

checkpoint = 'transfo-xl-wt103'
revision = '40a186da79458c9f9de846edfaea79c412137f97'

tokenizer = TransfoXLTokenizer.from_pretrained(checkpoint, revision=revision)
model = TransfoXLLMHeadModel.from_pretrained(checkpoint, revision=revision, torchscript=True)

text = "Hello, my dog is cute"
encoded_input = tokenizer(text, return_tensors='pt')
example_inputs = dict(encoded_input)

s_params = inspect.signature(model.forward).parameters
input_parameters, _, model, _ = prepare_example_inputs_and_model(example_inputs, s_params, model)

model.eval()
tm = torch.jit.trace(model, **input_parameters)
print(tm.inlined_graph)

om = ov.convert_model(model, example_input=example_inputs)
```

This model also has issue with `aten::add` for boolean inputs, but that is covered by a different ticket: #21971

- Implement conversion rule and/or transformation to support the new operation.
- Implement operation tests in tests/layer_tests/pytorch_tests. Please consider different data types, but keep reasonable number of test cases

### Example Pull Requests

Good place to start is `aten::index_put_` operation as reference for `aten::index_copy_`

### Resources

- [Contribution guide - start here!](https://github.com/openvinotoolkit/openvino/blob/master/CONTRIBUTING.md)
- [What is OpenVINO?](https://github.com/openvinotoolkit/openvino#what-is-openvino-toolkit)
- [OpenVINO PyTorch Frontend](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/pytorch/README.md)
- [Blog post on contributing to OpenVINO](https://github.com/openvinotoolkit/openvino/blob/master/CONTRIBUTING.md)
- [User documentation](https://docs.openvino.ai/)

### Contact points

@openvinotoolkit/openvino-pytorch-frontend-maintainers

### Ticket

_No response_

### PR DESCRIPTION
**Pull request by AsakusaRinne**
### Details:
 - *Add support for aten::index_copy_ operation*
 - *Add related unit test of it*

### Tickets:
 - *No ticket*

### Related issues
- Closes #21970

### CODE REVIEW

**Comment by AsakusaRinne** [comment]
@mvafin Could you please help to review this PR which resolves #21970? I've really had a hard time working on it and spent more than 8 hours to figure out all the things. Three other developers before me took the issue and ended with giving up. I'm happy that finally I make it! 😀
@ravi9 I'm a developer applying for your GSoC project. The first two issues I've done are related to TF frontend and the batch plugin. So I took this issue to gain experience about PyTorch frontend and adding aten operations. I'll appreciate it if you'd like to review it together. 😺

I test the code in #21970 to run `transfo-xl-wt103` and there's no error now.

Besides there're some questions I want to ask about:
1. At the beginning I referred to the implementation of `index_put_`, and found there's a check for whether `indices` is static. However, if I add the same check for `index_copy_`, an error `aten::index_copy_: dynamic rank for indices is not supported` will raise when compiling `transfo-xl-wt103`. Should I keep the check of it? I'm a bit confused of this behavior.
2. Some ops requires scalar as parameters, for example, `Range`. However, when I pass a tensor with shape `[1]` to it, the check will fail, saying it must be a scalar. I'm curious about why this conversion is not done implicitly in openvino.

File: src/frontends/pytorch/src/transforms/aten_index_copy_replacer.cpp [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+        /*
+         * Here we use an example to express the shape for each step
+         * input = torch.ranint([4, 3, 5])
+         * dim = 1
+         * index = torch.tensor([1, 0])
+         * tensor = torch.randint([4, 2, 5])
+         */
+
+        auto dim_scalar = std::dynamic_pointer_cast<v0::Constant>(dim.get_node_shared_ptr())->cast_vector<int>()[0];
```
Comment (mvafin) [inline src/frontends/pytorch/src/transforms/aten_index_copy_replacer.cpp:117]: Please verify that dim is indeed cons before calling `cast_vector`. Also please verify that vector has exactly 1 element before calling `[0]`

**Review by mvafin — CHANGES_REQUESTED** [review]
@AsakusaRinne First of all, thanks a lot for implementing conversion of this operation! 

> At the beginning I referred to the implementation of index_put_, and found there's a check for whether indices is static. However, if I add the same check for index_copy_, an error aten::index_copy_: dynamic rank for indices is not supported will raise when compiling transfo-xl-wt103. Should I keep the check of it? I'm a bit confused of this behavior.

We needed this verification in `aten::index_put_` because we had to know the first dimension of input in order to do conversion:
https://github.com/openvinotoolkit/openvino/blob/baf1c9d346b6d698fa76618d8a88935ce4dfb0c5/src/frontends/pytorch/src/transforms/aten_index_put_replacer.cpp#L85
Your implementation doesn't rely on this fact so you do not need to have this verification. The reason why `aten::index_put_` was done as transformation is mainly because of that requirement to have a static dimension known, so we had to be sure we converted as much operations as possible before converting it BUT here is the question to you: can you do `aten::index_copy_` conversion regularly in `src/frontends/pytorch/src/op/index_copy_.cpp` instead of using a transformation? My understanding is yes you can, so if you agree, please do it there.

> Some ops require a scalar as parameter, for example, Range. However, when I pass a tensor with shape [1] to it, the check will fail, saying it has to be a scalar. I'm curious about why this conversion is not done implicitly in openvino.

This is just a requirement of [spec](https://docs.openvino.ai/2023.3/openvino_docs_ops_generation_Range_4.html) for `Range` op.

> Is it not allowed to use dynamic control flow in operation translation? When implementing index_copy_, I had some cases that could be easily solved by if control but I failed to find such operation. I guess maybe it's the same with torch.fx, which also doesn't allow dynamic control flow (except dynamic shape). A dynamic control flow will break the generation of graph when compiling the model. Please correct me if I misunderstand it.

You can use dynamic control flow (operations like [Loop](https://docs.openvino.ai/2023.3/openvino_docs_ops_infrastructure_Loop_5.html) or [If](https://docs.openvino.ai/2023.3/openvino_docs_ops_infrastructure_If_8.html)), but it is better to avoid it if possible, because those operations have limited support in some plugins and are generally slower, but there are examples of using loops, for example https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/pytorch/src/op/gcd.cpp

File: src/frontends/pytorch/src/transforms/aten_index_copy_replacer.cpp (hunk 1 above)
Comment (mvafin) [reply src/frontends/pytorch/src/transforms/aten_index_copy_replacer.cpp:117]: This will not be required if done not as transformation.

**Comment by AsakusaRinne** [comment]
@mvafin Hi, thank you so much for your detailed reply!

> Here is the question to you: can you do aten::index_copy_ conversion regularly in src/frontends/pytorch/src/op/index_copy_.cpp instead of using a transformation? My understanding is yes you can, so if you agree, please do it there.

I'll do it in this way if that's a better way to implement it. Would you mind a delay of no more than a week of it?

> we had to be sure we converted as much operations as possible before converting it.

If I 'm not misunderstanding it, ops implemented directly are converted first and ops by transformation will be converted later. Could you please correct me if I'm wrong?

**Comment by mvafin** [comment]
> I'll do it in this way if that's a better way to implement it. Would you mind a delay of no more than a week of it? I had a fever last week and now I'm hurrying to write the application of gsoc. :)

That shouldn't be a problem, but your PR may be closed in 2 weeks if no activity.

> If I 'm not misunderstanding it, ops implemented directly are converted first and ops by transformation will be converted later. Could you please correct me if I'm wrong?

That is right. First we convert all ops that have conversion rules and other operations are left as `PtFrameworkNode` so we can not run shape and type inference through them, because those are generic ops. Transformations applied after all regular conversions are done and they are applied in order they are initialized.

**Comment by AsakusaRinne** [comment]
@mvafin Could you please review this PR again? I've switched to the regular way to implement this op. :)

**Comment by mvafin** [comment]
build_jenkins

File: src/frontends/pytorch/src/op/index_copy_.cpp [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    auto input_shape = context.mark_node(std::make_shared<v3::ShapeOf>(input, element::i32));  // [4, 3, 5]
+    auto selected_dim = context.mark_node(std::make_shared<v8::Gather>(input_shape, positive_dim, const_0));  // 3
+    auto selected_dim_correct_type = context.mark_node(std::make_shared<v1::ConvertLike>(selected_dim, index));
+    Output<Node> correct_index = context.mark_node(std::make_shared<v1::Add>(index, selected_dim_correct_type));
+    correct_index = context.mark_node(std::make_shared<v1::Mod>(correct_index, selected_dim_correct_type));
+    auto unsqueezed_index =
+        context.mark_node(std::make_shared<v0::Unsqueeze>(correct_index, const_neg_1));  // [[1], [0]]
+
+    // begin the computation
+    if (dim_scalar == 0) {
```
Comment (mvafin) [inline src/frontends/pytorch/src/op/index_copy_.cpp:128]: Can't you use [ScatterElementsUpdate](https://docs.openvino.ai/2023.3/openvino_docs_ops_movement_ScatterElementsUpdate_12.html) which actually support negative axis and negative indices?

**Review by mvafin — CHANGES_REQUESTED** [review]

File: src/frontends/pytorch/src/op/index_copy_.cpp (hunk 2 above)
Comment (AsakusaRinne) [reply src/frontends/pytorch/src/op/index_copy_.cpp:128]: Oh no......It seems that I just need to call this operation to complete this feature, without such complex transforms... I spent more than 8 hours on the coding of this issue, so sad😿

File: src/frontends/pytorch/src/op/index_copy_.cpp [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+OutputVector translate_index_copy_(const NodeContext& context) {
+    // aten::index_copy_(self, dim, index, tensor) → Tensor
+    num_inputs_check(context, 4, 4);
+    auto input = context.get_input(0);
+    auto dim = context.get_input(1);
+    auto index = context.get_input(2);
+    auto tensor = context.get_input(3);
+
+    auto result = context.mark_node(std::make_shared<v3::ScatterElementsUpdate>(input, index, tensor, dim));
```
Comment (mvafin) [inline src/frontends/pytorch/src/op/index_copy_.cpp:25]: I don't think it will be as easy as that, please look here, it seems to be a similar op https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/pytorch/src/op/index_add.cpp

Also consider the case when indexes collide

File: src/frontends/pytorch/src/op/index_copy_.cpp (hunk 2 above)
Comment (AsakusaRinne) [reply src/frontends/pytorch/src/op/index_copy_.cpp:124]: Hi, I found that in PyTorch, negative axis is not allowed now. Evidence could be found [here](https://github.com/pytorch/pytorch/blob/main/test/test_torch.py#L3452). Do you think I should keep it just the same as PyTorch, or support negative axis?

Comment (mvafin) [reply src/frontends/pytorch/src/op/index_copy_.cpp:124]: If pytorch doesn't support negative axis it is not needed to support them in conversion.

Comment (AsakusaRinne) [reply src/frontends/pytorch/src/op/index_copy_.cpp:124]: Without the concern for negative axis support, is it okay to keep using `ScatterNDUpdate`?

Comment (mvafin) [reply src/frontends/pytorch/src/op/index_copy_.cpp:124]: What about negative indices? If `ScatterElementsUpdate` simplifies the implementation and makes it because of that more optimal I would suggest using it.

Comment (AsakusaRinne) [reply src/frontends/pytorch/src/op/index_copy_.cpp]: You're right, using `ScatterElementsUpdate` makes things much easier! I have switched the implementation and is now ready for your review. I should discuss with you earlier for that, thus this PR will be completed much more quickly. Anyway, it's a good experience of working on this issue. :)

File: src/frontends/pytorch/src/op/index_copy_.cpp [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    auto const_1_vec = v0::Constant::create(element::i32, Shape{1}, {1});
+
+    Output<Node> tensor_rank = std::get<1>(get_shape_rank(context, tensor, true));
+    auto tensor_rank_correct_type = context.mark_node(std::make_shared<v1::ConvertLike>(tensor_rank, dim));
+    auto positive_dim = normalize_axis(context, dim, tensor_rank_correct_type);
+
+    // begin the computation
+    auto tensor_shape = context.mark_node(std::make_shared<v3::ShapeOf>(tensor, element::i32));
+    auto dim_vec = context.mark_node(std::make_shared<v1::Reshape>(positive_dim, const_1_vec, false));
+    auto broadcasted_index = context.mark_node(std::make_shared<v1::Broadcast>(index, tensor_shape, dim_vec));
```
[the remark this answers is not in the document]
Comment (AsakusaRinne) [reply src/frontends/pytorch/src/op/index_copy_.cpp:43]: Yes, I think so. For example, tensor_shape is `[3, 4, 5, 3]`, index has shape `[3]` and the axis is `0`. If we do not broadcast it explicitly with axis, it will be equal to broadcasting `[1, 1, 1, 3]` to `[3, 4, 5, 3]`, while what is desired is broadcasting `[3, 1, 1, 1]` to `[3, 4, 5, 3]`.

**Review by mvafin — APPROVED** [review]

**Comment by mvafin** [comment]
build_jenkins
